www/data_service/src/services/data/data.service.coffee

```diff
@@ -0,0 +1,146 @@
+# TODO rename this, or the dataService in waterfall view
+class Data extends Provider
+    constructor: ->
+    # TODO caching
+    cache: false
+
+    $get: ($log, $injector, $q, restService, socketService, helperService, ENDPOINTS) ->
+        return new class DataService
+            self = null
+            constructor: ->
+                self = @
+                socketService.eventStream.onUnsubscribe = @unsubscribeListener
+                @generateEndpoints()
+                # map of path: count
+                @listeners = {}
+
+            # generate functions for root endpoints
+            generateEndpoints: ->
+                ENDPOINTS.forEach (e) =>
+                    # capitalize endpoint names
+                    E = helperService.capitalize(e)
+                    @["get#{E}"] = (args...) =>
+                        @get(e, args...)
+
+            # the arguments are in this order: endpoint, id, child, id of child, query
+            get: (args...) ->
+                # keep defined arguments
+                args = args.filter (e) -> e?
+
+                # get the query parameters
+                [..., last] = args
+                if angular.isObject(last)
+                    query = args.pop()
+
+                # up to date array, this will be returned
+                updating = []
+
+                promise = $q (resolve, reject) =>
+
+                    # get the data from the rest api
+                    restPath = helperService.restPath(args)
+                    restPromise = restService.get(restPath, query)
+
+                    # start consuming WebSocket messages
+                    socketPath = helperService.socketPath(args)
+                    socketPromise = @startConsuming(socketPath)
+
+                    # TODO needs testing
+                    # store all messages before the classes subscribe for changes
+                    # resend once those are ready
+                    messages = []
+                    unsubscribe = socketService.eventStream.subscribe (data) ->
+                        messages.push(data)
+
+                    # wait for the promises to be resolved
+                    $q.all([restPromise, socketPromise]).then ([response]) =>
```

Comment:
you have to wait for the socket promise before starting the rest. Or you have race conditions.